src/codecs/png.rs

```diff
@@ -287,6 +292,16 @@ impl<'a, R: 'a + Read> ImageDecoder<'a> for PngDecoder<R> {
         let width = self.reader.info().width;
         self.reader.output_line_size(width) as u64
     }
+
+    fn set_limits(&mut self, limits: Limits) -> ImageResult<()> {
+        limits.check_support(&crate::io::LimitSupport::default())?;
+        let info = self.reader.info();
+        limits.check_dimensions(info.width, info.height)?;
+        self.limits = limits;
+        // TODO: add `png::Reader::change_limits()` and call it here
```

Comment:
We can't really implement a `change_limits` method on the `png` side.

The problem is that many of the allocations in that crate happen when you create the `png::Reader` by calling `png::Decoder::read_info` because that's where the image metadata is read and saved into the inner Info struct. Ironically, the crate allocates relatively little memory for decoding pixel data, because it is quite good at decoding that data in a streaming fashion.

Rather, the current plan is to switch to calling `Decoder::set_limits` in this method, and only later call `read_info` when the user queries metadata beyond what's stored in the IHDR.

An alternative possibility would be for the `png` to actually not store metadata chunks as it reads them, but rather to store the offsets of those chunks, and then `Seek` past them. This would bound the memory used by the decoder while still enabling the metadata to be loaded later. In that case, the limits could safely be set later. Adding the `Seek` bound would, however, be a breaking change both here and in the png crate.  (An added complication is that `BufReader`'s `seek()` implementation has some performance gotchas. It should be possible to work around them once [`Seek::seek_relative`](https://doc.rust-lang.org/beta/std/io/trait.Seek.html#method.seek_relative) is stabilized, but I don't know if/when that'll happen.)

Reply:
The idea is that `change_limits()` would look at the amount of memory already allocated by the `png` crate and report an error if the crate is already over budget. This would let us easily bail out of the total memory usage limit is exceeded (including all the metadata).

I don't think the `Seek` bound is a good idea (too limiting, largely defeats streaming decoding use cases) but a refactor to move the compositing logic to `png` and track the allocations there could be viable.